### ISSUE TITLE
[BUG] When calling ScreenshotAsync buffer never leaving memory

### ISSUE THREAD

**Issue by chaseaucoin** [microsoft/playwright-dotnet#2292]
**Context:**
- Playwright Version: [what Playwright version do you use?]
Tried every version from 1.14 to 1.25
- Operating System: [e.g. Windows, Linux or Mac]
Linux and Windows
- .NET version: [e.g. .NET 6]
.NET 6
- Browser: [e.g. All, Chromium, Firefox, WebKit]
All
- Extra: [any specific details about your environment]

**Code Snippet**

Help us help you! Put down a short code snippet that illustrates your bug and
that we can run and debug locally. For example:

```cs
using Microsoft.Playwright;
using System.Diagnostics;

while (true)
{   
    await DoRun();
}

async Task DoRun()
{
    using var playwright = await Playwright.CreateAsync();
    await using var browser = await playwright.Chromium.LaunchAsync(new() { Headless = true });
    await using var context = await browser.NewContextAsync();
    var page = await context.NewPageAsync();

    for (int i = 0; i < 500; i++)
    {
        if (i % 10 == 0)
        {
            long memoryUsed = Process.GetCurrentProcess().WorkingSet64;
            Console.WriteLine($"{memoryUsed / 1024 / 1024} MB");
        }

        await page.GotoAsync("https://playwright.dev/dotnet");
        await page.ScreenshotAsync(new() { Path = "screenshot.png", FullPage = true });
    }
}
```

**Describe the bug**
The large number of iterations is to show the issue. 

If you set the path for an image the bytes still get allocated to the .NET process instead of just going to the file so that it can then be streamed and better memory managed. 

Once the memory has been allocated it never deallocates even if you force a garbage collection. Notice the way I have it, it has a completely isolated scope so it should be deallocating but my guess is there is something static somewhere keeping the bytes alive forever. 

This make it difficult to have long running processes. 

Add any other details about the problem here.

Happy to help if you can point me in a direction.

**Comment by mxschmitt** [comment]
This is known, that you need to re-create the context and it's page to release the memory. See here: https://github.com/microsoft/playwright/issues/6319

If you have a long running job, then re-creating the context e.g. after every 30 screenshots/page visits would fix it: 

```csharp
using Microsoft.Playwright;
using System.Diagnostics;

using var playwright = await Playwright.CreateAsync();
await using var browser = await playwright.Chromium.LaunchAsync(new() { Headless = true });
var context = await browser.NewContextAsync();
var page = await context.NewPageAsync();

for (int i = 0; i < 5000; i++)
{
    if (i % 10 == 0)
    {
        long memoryUsed = Process.GetCurrentProcess().WorkingSet64;
        Console.WriteLine($"{memoryUsed / 1024 / 1024} MB");
        await context.CloseAsync();
        context = await browser.NewContextAsync();
        page = await context.NewPageAsync();
    }

    await page.GotoAsync("https://playwright.dev/dotnet");
    await page.ScreenshotAsync(new() {  });
    Console.WriteLine($"Taking screenshot {i}");
}
await context.CloseAsync();
```

Folding into https://github.com/microsoft/playwright/issues/6319 since this is not a Playwright for .NET bug.

**Comment by chaseaucoin** [comment]
I'm not sure why you closed this when in my example this runs in a loop and thus disposes the browser context and playwright at the end of each cycle. 

This is also specific to images. If the screenshot is not made the memory gets released.

**Comment by mxschmitt** [comment]
Yup you dispose, but only after 500 navigations. A single navigation will keep track of all the requests and responses of the website you are navigating to. Around 2MB * 500 = a lot.

**Comment by chaseaucoin** [comment]
You'll see the same problem with 30 or 10 the volume of RAM being used will just be less but it never releases.

**Comment by chaseaucoin** [comment]
It is specific to images I've tested this fairly rigorously with combinations of set-up and tear down with images and not. If images are not collected the memory fully recovers on dispose, with images it does not.

**Comment by chaseaucoin** [comment]
MORE specifically when you specify the path the .NET process should have effectively 0 extra allocations to it's memory since it is calling the playwright server in node.

**Comment by mxschmitt** [comment]
I was able to repro it, sorry for closing this too fast! My bad. I can clearly see that memory is increasing on the .NET side.

**Comment by chaseaucoin** [comment]
Looking at the following code it looks let we are getting back the image as a JSON payload from the server. And then doing the save local to the .NET process. 

```cs
internal async Task<byte[]> ScreenshotAsync(string path, bool? omitBackground, ScreenshotType? type, int? quality, IEnumerable<ILocator> mask, ScreenshotAnimations? animations, ScreenshotCaret? caret, ScreenshotScale? scale, float? timeout)
{
    var args = new Dictionary<string, object>
    {
        ["type"] = type,
        ["omitBackground"] = omitBackground,
        ["path"] = path,
        ["timeout"] = timeout,
        ["animations"] = animations,
        ["caret"] = caret,
        ["scale"] = scale,
        ["quality"] = quality,
    };
    if (mask != null)
    {
        args["mask"] = mask.Select(locator => new Dictionary<string, object>
        {
            ["frame"] = ((Locator)locator)._frame._channel,
            ["selector"] = ((Locator)locator)._selector,
        }).ToArray();
    }

    return (await Connection.SendMessageToServerAsync(Guid, "screenshot", args).ConfigureAwait(false))?.GetProperty("binary").GetBytesFromBase64();
}
```

Would it be possible to treat it like an Artifact like Video does so that it could either be streamed into memory or streamed directly to disk? I think it is the unbuffered base 64 which is going to double up the memory usage for each image request. 

Its also very probable that the base64 string might also be where the memory is not being released.

```cs
internal class Video : IVideo
{
    private readonly TaskCompletionSource<Artifact> _artifactTcs = new();
    private readonly bool _isRemote;

    public Video(Page page, Connection connection)
    {
        _isRemote = connection.IsRemote;
        page.Close += (_, _) => _artifactTcs.TrySetCanceled();
        page.Crash += (_, _) => _artifactTcs.TrySetCanceled();
    }

    public async Task DeleteAsync()
    {
        var artifact = await _artifactTcs.Task.ConfigureAwait(false);
        await artifact.DeleteAsync().ConfigureAwait(false);
    }

    public async Task<string> PathAsync()
    {
        if (_isRemote)
        {
            throw new PlaywrightException("Path is not available when connecting remotely. Use SaveAsAsync() to save a local copy.");
        }
        var artifact = await _artifactTcs.Task.ConfigureAwait(false);
        return artifact.AbsolutePath;
    }

    public async Task SaveAsAsync(string path)
    {
        var artifact = await _artifactTcs.Task.ConfigureAwait(false);
        await artifact.SaveAsAsync(path).ConfigureAwait(false);
    }

    internal void ArtifactReady(Artifact artifact) => _artifactTcs.TrySetResult(artifact);
}
```

**Comment by chaseaucoin** [comment]
So this may all be related to the other bug doing some memory diagnostics it *appears* that the ConnectionCallback concurrent dictionaries are rooting a bunch of strings, byte arrays, and objects and never letting go of them.

**Comment by chaseaucoin** [comment]
But the other bug seems to be related to memory issues with playwright itself on the server. These issues are specific to the .NET implementation of the client.

**Comment by mxschmitt** [comment]
It seems like in this issue we make more assumptions than we have actual evidence that things go wrong. I fixed recently an issue with connection callbacks, it gets released in 1.26, but this should not be very significant and be relevant to this reproduction. I ran [the repro](https://github.com/microsoft/playwright-dotnet/issues/2292#issuecomment-1237266630) for 10 minutes and the memory is always between 45-60 MB which looks expected.

Could you elaborate how long you run it and what your memory expectation is? And maybe share screenshots / more diagnose relevant logs. This would help a lot. Thanks! Just wanted to organize it a bit, since it seems we are jumping around between different issues.

**Comment by mxschmitt** [comment]
Closing as part of the triage process since it seemed stale. Please create a new issue with a detailed reproducible or feature request if you still face issues.
